File: Medium/2347-count-nodes-equal-to-average-of-subtree/count-nodes-equal-to-average-of-subtree.py
# Definition for a binary tree node.
# class TreeNode:
#     def __init__(self, val=0, left=None, right=None):
#         self.val = val
#         self.left = left
#         self.right = right
class Solution:
    def solve(self, root, count, res):
        if root is None:
            return 0

        lcount=[0]
        rcount=[0]
        lsum=self.solve(root.left, lcount, res)
        rsum=self.solve(root.right, rcount, res)

        sum=root.val+lsum+rsum
        count[0]=1+lcount[0]+rcount[0]

        if root.val == sum//count[0]:
            res[0] += 1

        return sum
    
    def averageOfSubtree(self, root: TreeNode) -> int:
        res=[0]
        count=[0]

        self.solve(root, count, res)
        return res[0]

File: Medium/2347-count-nodes-equal-to-average-of-subtree/count-nodes-equal-to-average-of-subtree.java
/**
 * Definition for a binary tree node.
 * public class TreeNode {
 *     int val;
 *     TreeNode left;
 *     TreeNode right;
 *     TreeNode() {}
 *     TreeNode(int val) { this.val = val; }
 *     TreeNode(int val, TreeNode left, TreeNode right) {
 *         this.val = val;
 *         this.left = left;
 *         this.right = right;
 *     }
 * }
 */
import java.util.concurrent.atomic.AtomicInteger;

class Solution{
    int solve(TreeNode root, AtomicInteger count, AtomicInteger res){
        if(root == null)
            return 0;

        AtomicInteger lcount = new AtomicInteger(0); 
        AtomicInteger rcount = new AtomicInteger(0);
        int lsum=solve(root.left, lcount, res);
        int rsum=solve(root.right, rcount, res);

        int sum=root.val+lsum+rsum;
        count.set(1+lcount.get()+rcount.get());

        if(root.val == sum/count.get())
            res.incrementAndGet();

        return sum;
    }

    public int averageOfSubtree(TreeNode root){
        AtomicInteger res=new AtomicInteger(0);
        AtomicInteger count=new AtomicInteger(0);

        solve(root, count, res);
        return res.get();

    }
}




















